### PR TITLE
[WIP] dialects: Add more stencil verification

### PR DESCRIPTION
**Pull request by PapyChacal**
Add more verifiers to ApplyOp (matching bounds and matchin return types)
Add Traits on ReturnOp (HasParent(ApplyOp) nd IsTerminator)
Add a test for a StencilType parser corner case.

### CODE REVIEW

**Review by AntonLydike — APPROVED** [review]
:guitar:

File: tests/filecheck/dialects/stencil/invalid.mlir [hunk 1]
```diff
@@ -105,3 +115,55 @@ builtin.module {
...
+    ^0(%intb : !stencil.temp<?xf64>):
+      %v = "stencil.access"(%intb) {"offset" = #stencil.index<-1>} : (!stencil.temp<?xf64>) -> f64
+      "stencil.return"(%v, %v) : (f64, f64) -> ()
+    }) : (!stencil.temp<?xf64>) -> (!stencil.temp<?xf64>, !stencil.temp<[0,64]xf64>)
+    "stencil.store"(%outt1, %out) {"lb" = #stencil.index<0>, "ub" = #stencil.index<68>} : (!stencil.temp<?xf64>, !stencil.field<[-4,68]xf64>) -> ()
+    "func.return"() : () -> ()
+  }
+}
+
+// CHECK: Expected all output types bounds to be equals.
```
Comment (superlopuh) [inline tests/filecheck/dialects/stencil/invalid.mlir]:
```suggestion
// CHECK: Expected all output types bounds to be equals.

```

Comment (PapyChacal) [reply tests/filecheck/dialects/stencil/invalid.mlir]: :scream: 
Weren't those supposed to be catched by CI now? I might push a lot of those otherwise... I'm amazed how often I forget them. Reminds me of the early days' missing `;`s in C..

Comment (superlopuh) [reply tests/filecheck/dialects/stencil/invalid.mlir]: You might want to install the pre-commit hook

Comment (PapyChacal) [reply tests/filecheck/dialects/stencil/invalid.mlir]: Oh, that's what.. thanks!
